### DISCUSSION TITLE
User 클래스 직접 사용 vs. get_user_model로 동적 연결

### DISCUSSION THREAD

**Discussion by cbscsm** [8percent/styleguide#48] [General]
- [타입 어노테이션 관련 PR](https://github.com/8percent/eight/pull/16010)에서 일부 논의가 된 내용입니다.

---

[배경]
- Django에서는 "재사용 가능한 앱"을 만드는 경우에 유저 모델을 정적으로 지정하지 말고 `get_user_mdoel` 함수를 통해 동적으로 `settings`로 부터 가져오길 권장하고 있습니다.
- 이에 따라 저희도 `get_user_mdoel`을 사용하고 있습니다.

[제약]
- `get_user_model`에서 리턴되는 클래스가 어떤 기능성을 가지는 지 정적으로 정의되지 못합니다. 이 이유로, 타입 어노테이션에 사용될 수 없습니다.
- 타입힌트를 위해서는
  - 각 앱별 프로토콜(정적 덕 타이핑)을 정의해서 사용하거나
  - 최상위 추상 유저 모델을 두고 그것을 사용하거나
  - **실제 사용되는 유저 모델을 직접 지정해야 합니다.**

[제안 배경]
- 지금 타입 어노테이션을 위해서는 세 번째 방법을 사용하고 있습니다. 만약 단일 파일에서 런타임/타입체커 모두에게 User 모델이 필요한 경우 alias 해서 둘 다 정의해야합니다.
- 아직 "재사용 가능한 앱"을 고려해서 개발을 하고 있지 않고, 앞으로도 "두 개 이상"의 유저 모델이 생길 가능성이 거의 없어 보입니다.

---

[제안]
- 이에 따라, 모든 코드에서 `get_user_model` 대신 `apps.users.models.User` 클래스를 직접 사용하는 것을 제안합니다.

[장점]
- 정적/동적 대상 클래스 간 괴리가 생기지 않습니다.
- 코드가 더 추적/예측하기 쉬워집니다.

[단점]
- 추후 하나의 앱을 두 개 이상의 유저 모델에 사용해야하는 경우 많은 수정이 필요할 수 있습니다.

**Comment by Kirade** [comment]
get_user_model 을 통해 동적으로 지정된 사용자 모델을 활용하는 코드가 미래에 작성될 가능성이 적다고 생각합니다.
그래서 직접 import 하는 방식으로 활용하는것에 동의합니다.
